### ISSUE TITLE
Only One Instance of the Thermostat works as Expected

### ISSUE THREAD

**Issue by dstewartwhite** [ScratMan/HASmartThermostat#133]
[Smart Thermostat PID.pdf](https://github.com/ScratMan/HASmartThermostat/files/10786638/Smart.Thermostat.PID.pdf)
[HA Config Text.txt](https://github.com/ScratMan/HASmartThermostat/files/10786642/HA.Config.Text.txt)
[home-assistant_2023-02-20T16-25-48.406Z.log](https://github.com/ScratMan/HASmartThermostat/files/10786644/home-assistant_2023-02-20T16-25-48.406Z.log)

**Describe the bug**
There are two configurations of the Thermostat in the configuration.yaml file. Only the first configuration thermostat works as expected. The second does not trigger any response to  an actual temperature difference to the Target. 

**To Reproduce**
Create two configurations to control separate heaters at different temperatures

**Expected behaviour**
That both configurations would control their associated heaters individually.

**Screenshots**
![image](https://user-images.githubusercontent.com/114423605/220171553-bb40c0dd-23e1-4b1f-b029-e4de6b83d421.png)

Shows the heater being turn Off despite a temperature difference after being triggered on manually 
![image](https://user-images.githubusercontent.com/114423605/220171233-0b4c5df2-fc80-4948-9743-e9edc2d6a467.png)

**Desktop (please complete the following information):**
 - OS: Debian11
 - Browser Chrome
 - Version Home Assistant 2023.2.5

**Additional context**
If each configuration is entered into the configuration.yaml file separately they work as intended controlling the specified heater. If the second heater is triggered manually, irrespective of the temperature error the thermostat switches Off the heater.

**Comment by ScratMan** [comment]
Hi, which version are you using ?

**Comment by dstewartwhite** [comment]
Hi

Upgraded to your latest this morning 2023.2.1 prior to submitting the Issue #133.

Regards

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by ScratMan** [comment]
Your log only mentions some OFF commands, the control output is always 0.00 because temperature is above the set point on both thermostats, so there is no activation needed, and the switches are never toggled ON.

Your capture seems to be more recent than the log, and the switch off is requested by home_assistant.turn_off, so it seems to be related to the smart thermostat, bit the switch on is requested by the service switch.turn_on ; so it looks like something (an automation or a script) is turning on the switch and the keep_alive of the smart thermostat then forces the switch to OFF as there is no need for heating.

**Comment by dstewartwhite** [comment]
Attached a new log file to ensure all actions are in sequence
[home-assistant_2023-02-21T09-58-11.708Z.log](https://github.com/ScratMan/HASmartThermostat/files/10792252/home-assistant_2023-02-21T09-58-11.708Z.log)

This screen dump shows that the Seed thermostat (2nd in Config file) target temp is above the target sensor temp, so the thermostat should I believe be activated and indicate "Heat" not "Idle"
![image](https://user-images.githubusercontent.com/114423605/220312747-77487a56-e622-4243-b70c-9d367b1d3254.png)

Log book of the seed heater switch. Not turned on after adding second thermostat to config and restart.
![image](https://user-images.githubusercontent.com/114423605/220314043-484ae09a-aef3-4a16-978b-196c27f011ef.png)

I disabled the automation scripts that are used  until I can get the second thermostat to work.

The final screen dump shows the history of the seed thermostat from the last HA restart
![image](https://user-images.githubusercontent.com/114423605/220316270-f8f725c5-681d-402b-ab84-da743c90400a.png)

**Comment by ScratMan** [comment]
That's strange as the log mentions output of 0.00 with error=2.10 and p=0.00 while the kp is 40.

Could you please set debug: true for both thermostats in the config and post the full state of each thermostat by going to Development tools \ States and selecting them in the list.
The "entity state" Yaml should look like this :
```yaml
hvac_modes:
  - heat
  - 'off'
min_temp: 7
max_temp: 30
preset_modes:
  - none
  - away
  - eco
  - boost
  - comfort
current_temperature: 21.1
temperature: 21
hvac_action: idle
preset_mode: none
away_temp: 14
eco_temp: 19
boost_temp: 25
comfort_temp: 21
home_temp: null
sleep_temp: null
activity_temp: null
control_output: 0
kp: 64
ki: 0.007
kd: 125000
ke: 0.6
pid_mode: auto
pid_i: 2.5
friendly_name: Thermostat salle de bain
supported_features: 17

```

**Comment by dstewartwhite** [comment]
As requested, both thermostat's debug set to "true" and HA restarted

Entity States

hvac_modes:
  - heat
  - 'off'
min_temp: 7
max_temp: 35
target_temp_step: 0.1
current_temperature: 19.5
temperature: 20
hvac_action: idle
away_temp: null
eco_temp: null
boost_temp: null
comfort_temp: null
home_temp: null
sleep_temp: null
activity_temp: null
control_output: 0
kp: 0
ki: 0
kd: 0
ke: 0
pid_mode: auto
pid_i: 0
friendly_name: Smart Thermostat Seed (PID)
supported_features: 1

hvac_modes:
  - heat
  - 'off'
min_temp: -10
max_temp: 40
target_temp_step: 0.1
current_temperature: 16.8
temperature: 10
hvac_action: idle
away_temp: 14
eco_temp: null
boost_temp: null
comfort_temp: null
home_temp: null
sleep_temp: null
activity_temp: null
control_output: 0
kp: 5
ki: 0.01
kd: 500
ke: 0
pid_mode: auto
pid_i: 2.7
friendly_name: Propagator Heating Control (PID)
supported_features: 1

**Comment by ScratMan** [comment]
Here is the root cause of your issue:

```yaml
kp: 0
ki: 0
kd: 0
ke: 0

```

The gains are restored from previous known state at boot, and the values in the YAML are ignored except is the HA database is deleted or crashes. The first release of 2023 introduced a bug that prevented correct restoration of the gains and it forced them to 0. That's why the output of the PID is always 0.

You need to use the `smart_thermostat.set_pid_gains` service to set back the gains to the expected values.

**Comment by dstewartwhite** [comment]
Hi
I have used the service to set the PID values and now it is working as expected. 

I have a couple of further comments.
The issue ONLY appears to effect the second configuration, in my case, as its gains values are being set to 0. The first configuration values are not effected.
The issue was present in at least the last two releases of 2022, as I first started to get two thermostats running in November last year.
Regards

**Comment by ScratMan** [comment]
There have been no release between 2022.2.1 in February '22 and 2023.1.0.
In the 2023.1.0 I brought back the integration reload service, but the timers were not stopped when the thermostat was unloaded by Home Assistant. So if something reloaded the integration, it created a second set of timers for the same entity, and depending on the sync between the redundant timers, it could create some big troubles. It also introduced a bug in the gains value restoration, that has been fixed in 2023.1.1.
The 2023.2.0 release fixed the timer issue, so when setting back the correct gains on latest release, everything should be fine again.
